contractcourt/contract_resolvers_test.go

```diff
@@ -1 +1,344 @@
+	signDesc *input.SignDescriptor) ([]byte, error) {
+	return nil, nil
+}
+
+func (m *mockSigner) ComputeInputScript(tx *wire.MsgTx,
+	signDesc *input.SignDescriptor) (*input.Script, error) {
+	return nil, nil
+}
+
+type mockWitnessBeacon struct {
+	preImageUpdates chan lntypes.Preimage
+
+	newPreimages chan []lntypes.Preimage
+}
+
+func (m *mockWitnessBeacon) SubscribeUpdates() *WitnessSubscription {
+	return &WitnessSubscription{
+		WitnessUpdates:     m.preImageUpdates,
+		CancelSubscription: func() {},
+	}
+}
+
+func (m *mockWitnessBeacon) LookupPreimage(payhash lntypes.Hash) (lntypes.Preimage, bool) {
+
+	return lntypes.Preimage{}, false
+}
+
+func (m *mockWitnessBeacon) AddPreimages(preimages ...lntypes.Preimage) error {
+	m.newPreimages <- preimages
+	return nil
+}
+
+// TestHtlcTimeoutResolver tests that the timeout resolver properly handles all
+// variations of possible local+remote spends.
+func TestHtlcTimeoutResolver(t *testing.T) {
+	t.Parallel()
+
+	fakePreimageBytes := bytes.Repeat([]byte{1}, lntypes.HashSize)
+
+	var (
+		htlcOutpoint wire.OutPoint
+		fakePreimage lntypes.Preimage
+	)
+	fakeSignDesc := &input.SignDescriptor{
+		Output: &wire.TxOut{},
+	}
+
+	copy(fakePreimage[:], fakePreimageBytes)
+
+	signer := &mockSigner{}
+	sweepTx := &wire.MsgTx{
+		TxIn: []*wire.TxIn{
+			{
+				PreviousOutPoint: htlcOutpoint,
+				Witness:          [][]byte{{0x01}},
+			},
+		},
+	}
+	fakeTimeout := int32(5)
+
+	templateTx := &wire.MsgTx{
+		TxIn: []*wire.TxIn{
+			{
+				PreviousOutPoint: htlcOutpoint,
+			},
+		},
+	}
+	genTimeoutTx := func() (*wire.MsgTx, error) {
+		witness, err := input.ReceiverHtlcSpendTimeout(
+			signer, fakeSignDesc, sweepTx,
+			fakeTimeout,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		templateTx.TxIn[0].Witness = witness
+		return templateTx, nil
+	}
+
+	testCases := []struct {
+		// name is a human readable
+		name string
+
+		// remoteCommit denotes if the commitment broadcast was the
+		// remote commitment or not.
+		remoteCommit bool
+
+		// timeout denotes if the HTLC should be let timeout, or if the
+		// "remote" party should sweep it on-chain. This also affects
+		// what type of resolution message we expect.
+		timeout bool
+
+		// txToBroadcast is a function closure that should generate the
+		// transaction that should spend the HTLC output. Test authors
+		// can use this to customize the witness used when spending to
+		// trigger various redemption cases.
+		txToBroadcast func() (*wire.MsgTx, error)
+	}{
+		// Remote commitment is broadcast, we time out the HTLC on
+		// chain, and should expect a fail HTLC resolution.
+		{
+			name:          "timeout remote tx",
+			remoteCommit:  true,
+			timeout:       true,
+			txToBroadcast: genTimeoutTx,
+		},
+
+		// Our local commitment is broadcast, we timeout the HTLC and
+		// still expect an HTLC fail resolution.
+		{
+			name:          "timeout local tx",
+			remoteCommit:  false,
+			timeout:       true,
+			txToBroadcast: genTimeoutTx,
```

Comment:
Shouldn't the tx here use this stack (timeout tx)?
`0 <remotehtlcsig> <localhtlcsig> 0`

Reply:
Nice catch, fixed!